include/libsemigroups/detail/path-iterators.hpp

```diff
@@ -0,0 +1,380 @@
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+// This file contains declarations of the iterator classes for
+// paths in an WordGraph.
+
+#ifndef LIBSEMIGROUPS_DETAIL_PATH_ITERATORS_HPP_
+#define LIBSEMIGROUPS_DETAIL_PATH_ITERATORS_HPP_
+
+namespace libsemigroups {
+  namespace detail {
+    ////////////////////////////////////////////////////////////////////////
+    // pilo = Path And (terminal) Node In Lex Order
+    ////////////////////////////////////////////////////////////////////////
+
+    template <typename Node>
+    class const_pilo_iterator {
+     public:
+      using node_type       = Node;
+      using label_type      = typename WordGraph<Node>::label_type;
+      using value_type      = word_type;
+      using size_type       = typename std::vector<value_type>::size_type;
+      using difference_type = typename std::vector<value_type>::difference_type;
+      using const_pointer   = typename std::vector<value_type>::const_pointer;
+      using pointer         = typename std::vector<value_type>::pointer;
+      using const_reference = typename std::vector<value_type>::const_reference;
+      using reference       = const_reference;
+      using iterator_category = std::forward_iterator_tag;
+
+     private:
+      value_type             _edges;
+      WordGraph<Node> const* _word_graph;
+      label_type             _edge;
+      size_t                 _min;
+      size_t                 _max;
+      std::vector<node_type> _nodes;
+
+     public:
+      const_pilo_iterator() noexcept;
+      const_pilo_iterator(const_pilo_iterator const&);
+      const_pilo_iterator(const_pilo_iterator&&) noexcept;
+      const_pilo_iterator& operator=(const_pilo_iterator const&);
+      const_pilo_iterator& operator=(const_pilo_iterator&&) noexcept;
+      ~const_pilo_iterator();
+
+      const_pilo_iterator(WordGraph<node_type> const* ptr,
+                          node_type                   source,
+                          size_type                   min,
+                          size_type                   max);
+
+      // noexcept because comparison of std::vector<node_type>'s is noexcept
+      // because comparison of node_type's is noexcept
+      [[nodiscard]] bool
+      operator==(const_pilo_iterator const& that) const noexcept {
+        return _nodes == that._nodes;
+      }
+
+      // noexcept because operator== is noexcept
+      [[nodiscard]] bool
+      operator!=(const_pilo_iterator const& that) const noexcept {
+        return !(this->operator==(that));
+      }
+
+      [[nodiscard]] const_reference operator*() const noexcept {
+        return _edges;
+      }
+
+      [[nodiscard]] const_pointer operator->() const noexcept {
+        return &_edges;
+      }
+
+      // TODO(1) to tpp
+      [[nodiscard]] node_type target() const noexcept {
+        if (!_nodes.empty()) {
+          return _nodes.back();
+        } else {
+          return UNDEFINED;
+        }
+      }
```

Comment:
Should the `TODO(1)` be done in this PR?

Reply:
Later